service/tablet_allocator.cc

```diff
@@ -1076,45 +1054,122 @@ class load_balancer {
         return {s, rs};
     }
 
+    struct table_sizing {
+        size_t current_tablet_count; // Tablet count in group0.
+        size_t target_tablet_count; // Tablet count wanted by scheduler.
+        std::optional<uint64_t> avg_tablet_size; // nullptr when stats not yet available.
+
+        size_t target_tablet_count_aligned; // target_tablet_count aligned to power of 2.
+        resize_decision::way_type resize_decision; // Decision which should be emitted to achieve target_tablet_count_aligned.
+    };
+
+    struct sizing_plan {
+        std::unordered_map<table_id, table_sizing> tables;
+    };
+
+    future<sizing_plan> make_sizing_plan() {
+        sizing_plan plan;
+
+        auto process_table = [&] (table_id table, schema_ptr s, const tablet_aware_replication_strategy* rs, size_t tablet_count) -> future<> {
+            table_sizing& table_plan = plan.tables[table];
+            table_plan.current_tablet_count = tablet_count;
+
+            size_t target_tablet_count = 1;
+
+            target_tablet_count = std::max<size_t>(target_tablet_count,
+                    co_await rs->calculate_min_tablet_count(s, _tm, _target_tablet_size, _initial_scale));
+
+            // The block below establishes table_plan.target_tablet_count
+
+            const auto* table_stats = load_stats_for_table(table);
+            if (table_stats) {
+                auto avg_tablet_size = table_stats->size_in_bytes / std::max<size_t>(table_plan.current_tablet_count, 1);
+                auto tablet_count_from_size = table_plan.current_tablet_count;
+
+                auto cur_decision = _tm->tablets().get_tablet_map(table).resize_decision();
+                // Split based on min_tablet_count or avg_tablet_size, or
+                // if the current resize_decision is split, apply hysteresis,
+                // so it would get cancelled only when crossing back the half-way point.
+                if (avg_tablet_size > target_max_tablet_size() ||
+                    (cur_decision.is_split() && avg_tablet_size >= _target_tablet_size)) {
+                    // TODO: extend to n-way split when needed
+                    tablet_count_from_size *= 2;
```

Comment:
This does much more than refactoring the existing code in to_resize_decision.
Is there a way to break this patch into two: first one does only refactoring and the following patch implementing the changes to the existing algorithm?

Reply:
> This does much more than refactoring the existing code in to_resize_decision, as the patch title suggests.
Is there a way to break this patch into two: first one does only refactoring and the following patch implementing the changes to the existing algorithm?

Reply:
I can do that.